SharedMemory/RpcBuffer.cs

```diff
@@ -745,7 +748,75 @@ protected virtual async Task<RpcResponse> SendMessage(MessageType msgType, RpcRe
             }
         }
 
-        bool WriteProtocolV1(MessageType msgType, ulong msgId, byte[] msg, ulong responseMsgId, int timeout)
+        // shamelessly copied from Stephen Cleary's https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Interop.WaitHandles/Interop/WaitHandleAsyncFactory.cs
+        // his nuget package does not target all the tfms we need
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+        public static Task<bool> FromWaitHandle(WaitHandle handle, TimeSpan timeout)
+        {
+            return FromWaitHandle(handle, timeout, CancellationToken.None);
+        }
+
+        public static Task<bool> FromWaitHandle(WaitHandle handle, CancellationToken token)
+        {
+            return FromWaitHandle(handle, Timeout.InfiniteTimeSpan, token);
+        }
+
+        public static class TaskConstants
+        {
+            public static readonly Task<bool> BooleanTrue = Task.FromResult(true);
+            public static readonly Task<bool> BooleanFalse = TaskConstants<bool>.Default;
+        }
+
+        public static class TaskConstants<T>
+        {
+            public static readonly Task<T> Default = Task.FromResult(default(T));
+        }
+
+        public static Task<bool> FromWaitHandle(WaitHandle handle, TimeSpan timeout, CancellationToken token)
```

Comment:
Why do you need all of the Task Results if you could make this async Task and return normal true or false or `return await DoFromWaitHandle(...)`?

Reply:
As the comments state, I copied this from Stephen Cleary's AsyncEx.
I assume he did this for performance reasons.
AFAIK, your proposed way would incur the `async` performance penalty unnecessarily just to return a constant value (unless the compiler is clever enough).
Also, reusing the same `Task` objects every time avoids allocation and helps with GC pressure as well.

But in the end, the fact that Stephen Cleary did it like this is good enough for me regardless.

Instead of copy-pasting his code, we could also add a package dependency on [Nito.AsyncEx.Interop.WaitHandles](https://www.nuget.org/packages/Nito.AsyncEx.Interop.WaitHandles/), which seems to be a small and lightweight package.
I did not do that initially because people were praising that SharedMemory has so few dependencies.

Reply:
Using async/await does incur a slight cost as a state machine is created, so using Task.FromResult makes sense.

@shuebner I would rather keep out any additional dependencies if possible. When deploying client/server applications it is nice to minimise what needs to come along for the ride (especially for some of the injection solutions this is used in - e.g. VR overlay). Referencing projects as sub-modules and compiling in directly is an option if absolutely necessary.